File: frontend/src/types/dem.ts
/**
 * DEM processing — mirrors `schemas/dem.py` (§6.2).
 *
 * ★ These types describe the THREE-STAGE algorithm in `gis/dem`, and nothing else:
 *   crop to an AOI, reproject to metres, sample elevations. There is no `fill_sinks`
 *   and no `smoothing` level — the scaffold offered both and the pipeline implements
 *   neither. A control that does nothing is the UI equivalent of a fabricated number.
 *
 * ★ Wire fields are snake_case and are never renamed (§8.1 / L9).
 */

/** A corner or point, EPSG:4326. Accepts decimal degrees or any DMS spelling. */
export interface DemPoint {
  /** `34.127544`, `34°07'39.16"N`, or `34; 7; 39.16; N`. */
  lat: string | number;
  /** Same accepted spellings. */
  lon: string | number;
}

/** Resampling kernels the backend offers. `bilinear` is the default for elevation. */
export type DemResampling =
  | 'nearest'
  | 'bilinear'
  | 'cubic'
  | 'cubic_spline'
  | 'lanczos'
  | 'average'
  | 'mode';

/** What stage 1 did. */
export interface DemCropSummary {
  source_crs: string;
  tolerance: number;
  /** `[width, height]` in pixels. */
  output_size: [number, number];
  /** Source pixel size, in the SOURCE CRS's units — degrees, for a geographic DEM. */
  pixel_size: [number, number];
  /**
   * ★ True when the padded AOI reached past the DEM's edge. The area of interest is
   * NOT fully covered by this tile, so part of it has no elevation at all.
   */
  clipped: boolean;
  aoi_corners_lonlat: [number, number][];
}

/** What stage 2 did — `(λ, φ, H) → (E, N, H)`. */
export interface DemReprojectSummary {
  source_crs: string;
  target_crs: string;
  resampling: string;
  /** True when the UTM zone was chosen from the DEM centre rather than given. */
  auto_utm: boolean;
  source_pixel_size: [number, number];
  /** ★ The DEM's ground sample distance, TRUE metres. The number worth reading. */
  output_pixel_size_m: [number, number];
  output_size: [number, number];
  output_bounds_m: [number, number, number, number];
}

/** Elevation range of the processed DEM. `null` when every cell is void. */
export interface DemStatistics {
  min_m: number | null;
  max_m: number | null;
  mean_m: number | null;
  valid_cells: number;
  void_cells: number;
}

/** The camera station an AOI was built from, and the zone it fixes. */
export interface DemCameraSummary {
  lat: number;
  lon: number;
  /** Working radius, TRUE ground metres. */
  radius_m: number;
  /**
   * ★ The UTM zone the camera position falls in — DERIVED, never typed. A zone entered
   * by hand can be wrong; one taken from where the camera stood cannot.
   */
  utm_epsg: string;
}

/** `POST /dem/process` → the full record of one run. */
export interface DemProcessResponse {
  run_id: string;
  /** Present when the AOI came from a camera position + radius. */
  camera: DemCameraSummary | null;
  source_name: string;
  source_crs: string;
  output_crs: string;
  output_size: [number, number];
  /** `null` when the output is not in a ground-metre CRS. */
  output_pixel_size_m: [number, number] | null;
  output_bytes: number;
  crop: DemCropSummary | null;
  reproject: DemReprojectSummary | null;
  statistics: DemStatistics;
  /** The AOI polygon, WGS84, ready to draw on a map. */
  aoi_geojson: Record<string, unknown> | null;
  /** Honest notes: a clamped crop, a skipped stage, a void-heavy output. */
  warnings: string[];
  download_url: string;
  /**
   * ★ True when this run became the DEM new GCPs sample elevation from. Existing GCPs
   * are NOT re-sampled — a recorded elevation is an observation, not a live query.
   */
  is_elevation_source: boolean;
}

/** The multipart form `POST /dem/process` accepts. */
/**
 * What the DEM pages hold after a pick: browser bytes, or a desktop PATH.
 *
 * ★ The path variant exists because the desktop shell and the API share one
 * machine: a 3.8 GB tile shipped as bytes must survive IPC + several in-memory
 * copies + an HTTP body (it didn't — files over 2 GiB failed outright), while a
 * path costs nothing and the backend streams straight from disk.
 */
export type DemSource =
  | { kind: 'file'; name: string; size: number; file: File }
  | { kind: 'path'; name: string; size: number; path: string };

export interface DemProcessForm {
  /** The upload's bytes (browser route). Exactly one of `file`/`source_path`. */
  file?: File;
  /** Absolute path on the API's own machine (desktop route). */
  source_path?: string;
  /** ★ Adopt the output as this project's elevation source (with
   *  `set_as_elevation_source: true`) — the processing page's project mode. */
  project_id?: string;
  /**
   * ★ Adopt the output for ONE image instead of the whole project (needs `project_id`
   * too). That image's GCPs and auto-GCP raycast then read THIS DEM; every other image
   * keeps the project DEM — the image-overrides-project rule.
   */
  image_id?: string;
  /** At least 3 corners, or omitted to process the whole DEM. */
  aoi_corners?: DemPoint[];
  /**
   * ★ Camera station. All three together or none. When given they define the AOI (a
   * disc of `radius_m` around the station) AND fix the UTM zone, so `target_srid` is
   * unnecessary — the zone follows from where the camera stood.
   */
  camera_lat?: number;
  camera_lon?: number;
  radius_m?: number;
  /** Padding fraction around the AOI. `0.10` is 10%. */
  tolerance?: number;
  reproject?: boolean;
  /** Explicit projected EPSG code, or omitted to auto-pick the UTM zone. */
  target_srid?: number | null;
  /**
   * ★ NOT SURFACED IN THE UI. The server defaults to `bilinear`, which is the right
   * kernel for a continuous elevation surface; the API keeps the parameter because it
   * is part of the algorithm, but the page does not ask.
   */
  resampling?: DemResampling;
  /** ★ NOT SURFACED. Omitted means "preserve the source resolution", the safe default. */
  target_resolution_m?: number | null;
  /** Adopt this DEM as the GCP elevation source. Defaults to true server-side. */
  set_as_elevation_source?: boolean;
}

/** One sampled point — stage 3's row. */
export interface DemSampledPoint {
  name: string;
  lat: number;
  lon: number;
  x: number | null;
  y: number | null;
  /** ★ `null` and `0` are different claims: null means void or off-grid. */
  dem_z_m: number | null;
  offset_m: number;
  z_m: number | null;
  inside_dem: boolean;
  /** Elevation outside the plausible land-surface range. Reported, never dropped. */
  suspicious: boolean;
}

/** `POST /dem/{run_id}/sample` body. */
export interface DemSampleForm {
  points: DemPoint[];
  names?: string[];
  method?: 'bilinear' | 'nearest';
  offset_m?: number;
  output_crs?: string | null;
}

/** `POST /dem/{run_id}/sample` result. */
export interface DemSampleResponse {
  run_id: string;
  dem_crs: string;
  output_crs: string;
  method: string;
  offset_m: number;
  points: DemSampledPoint[];
  inside_count: number;
  with_elevation_count: number;
}

/** `GET /dem/active` — which DEM currently feeds GCP elevations. */
export interface DemActiveResponse {
  active: boolean;
  /**
   * ★ True when `LE_ELEVATION_PROVIDER=dem_run`. A DEM can be adopted while the
   * provider is off, in which case nothing samples it — reported, not hidden.
   */
  provider_enabled: boolean;
  path?: string | null;
  /** ★ Set when the DEM belongs to ONE project; null for the server-wide fallback. */
  project_id?: string | null;
  /** ★ Set when the DEM belongs to ONE image (the per-image override); null otherwise. */
  image_id?: string | null;
  run_id?: string | null;
  source_name?: string | null;
  adopted_at?: string | null;
  output_crs?: string | null;
  output_pixel_size_m?: number[] | null;
  /** Reported as `elevation_ce90_m` on every GCP sampled from this DEM. */
  vertical_ce90_m?: number | null;
  /** `unknown` unless declared. Orthometric vs ellipsoidal differ by ~20 m here. */
  vertical_datum?: string | null;
  statistics?: Record<string, unknown> | null;
}

File: backend/app/schemas/dem.py
"""DEM processing — the wire contract for ``/dem`` (``gis.dem``).

Mirrors the three-stage algorithm's parameters exactly, and nothing else. The scaffold
page previously offered ``fill_sinks`` and a ``smoothing`` level; neither exists in the
algorithm, so neither exists here. Advertising a control the pipeline does not implement
is the same class of dishonesty as a fabricated confidence score.
"""

from __future__ import annotations

from typing import Annotated, Any, Literal

from pydantic import BaseModel, ConfigDict, Field, field_validator

__all__ = [
    "DemActiveResponse",
    "DemAoiCorner",
    "DemCameraSummary",
    "DemCropSummary",
    "DemProcessResponse",
    "DemReprojectSummary",
    "DemSampleRequest",
    "DemSampleResponse",
    "DemSampledPoint",
    "DemStatistics",
]


class DemAoiCorner(BaseModel):
    """One AOI corner. Accepts decimal degrees or any DMS spelling the algorithm reads."""

    model_config = ConfigDict(extra="forbid")

    lat: str | float = Field(
        ...,
        description=(
            "Latitude. Decimal degrees, or DMS such as `34°07'39.16\"N` or `34; 7; 39.16; N`."
        ),
        examples=["34°07'39.16\"N", 34.127544],
    )
    lon: str | float = Field(
        ...,
        description="Longitude, same accepted spellings as `lat`.",
        examples=["36°01'35.17\"E", 36.026436],
    )


class DemCropSummary(BaseModel):
    """What stage 1 did."""

    model_config = ConfigDict(extra="forbid")

    source_crs: str
    tolerance: float = Field(..., description="Padding fraction applied around the AOI.")
    output_size: tuple[int, int] = Field(..., description="(width, height) in pixels.")
    pixel_size: tuple[float, float] = Field(
        ..., description="Source pixel size, in the SOURCE CRS's units (degrees, usually)."
    )
    clipped: bool = Field(
        ...,
        description=(
            "True when the padded AOI reached past the DEM's edge and was clamped. "
            "The area of interest is not fully covered by this tile."
        ),
    )
    aoi_corners_lonlat: list[tuple[float, float]] = Field(default_factory=list)


class DemReprojectSummary(BaseModel):
    """What stage 2 did — ``(lambda, phi, H) -> (E, N, H)``."""

    model_config = ConfigDict(extra="forbid")

    source_crs: str
    target_crs: str
    resampling: str
    auto_utm: bool = Field(
        ..., description="True when the UTM zone was chosen from the DEM centre."
    )
    source_pixel_size: tuple[float, float]
    output_pixel_size_m: tuple[float, float] = Field(
        ...,
        description=(
            "★ The DEM's ground sample distance in TRUE metres — the number to read "
            "when judging what the surface can support."
        ),
    )
    output_size: tuple[int, int]
    output_bounds_m: tuple[float, float, float, float]


class DemStatistics(BaseModel):
    """Elevation range of the processed DEM. ``None`` when every cell is void."""

    model_config = ConfigDict(extra="forbid")

    min_m: float | None = None
    max_m: float | None = None
    mean_m: float | None = None
    valid_cells: int = 0
    void_cells: int = 0


class DemCameraSummary(BaseModel):
    """The camera station the AOI was built from, and the zone it fixes."""

    model_config = ConfigDict(extra="forbid")

    lat: float
    lon: float
    radius_m: float = Field(..., description="Working radius, TRUE ground metres.")
    utm_epsg: str = Field(
        ...,
        description=(
            "★ The UTM zone the camera position falls in. Derived, never typed — a zone "
            "entered by hand can be wrong; one taken from where the camera stood cannot."
        ),
    )


class DemProcessResponse(BaseModel):
    """``POST /dem/process`` — the full record of one run."""

    model_config = ConfigDict(extra="forbid")

    run_id: str = Field(..., description="Handle for download and sampling.")
    camera: DemCameraSummary | None = Field(
        None, description="Present when the AOI came from a camera position + radius."
    )
    source_name: str
    source_crs: str
    output_crs: str
    output_size: tuple[int, int]
    output_pixel_size_m: tuple[float, float] | None = Field(
        None,
        description="None when the output is not in a ground-metre CRS.",
    )
    output_bytes: int
    crop: DemCropSummary | None = None
    reproject: DemReprojectSummary | None = None
    statistics: DemStatistics
    aoi_geojson: dict[str, Any] | None = Field(
        None, description="The AOI polygon, WGS84, for drawing on the map."
    )
    warnings: list[str] = Field(
        default_factory=list,
        description="Honest notes: a clamped crop, a skipped stage, a void-heavy output.",
    )
    download_url: str
    is_elevation_source: bool = Field(
        False,
        description=(
            "True when this run's output was adopted as the DEM that new GCPs sample "
            "their elevation from. ★ Existing GCPs are NOT re-sampled — a recorded "
            "elevation is an observation, not a live query."
        ),
    )


class DemSampledPoint(BaseModel):
    """One sampled point — stage 3's row."""

    model_config = ConfigDict(extra="forbid")

    name: str
    lat: float
    lon: float
    x: float | None = Field(None, description="Easting in the output CRS, metres.")
    y: float | None = Field(None, description="Northing in the output CRS, metres.")
    dem_z_m: float | None = Field(
        None, description="Raw DEM elevation. ★ null and 0.0 are different claims."
    )
    offset_m: float = 0.0
    z_m: float | None = Field(None, description="dem_z_m + offset_m.")
    inside_dem: bool
    suspicious: bool = Field(
        False, description="Elevation outside the plausible land-surface range."
    )


class DemSampleRequest(BaseModel):
    """``POST /dem/{run_id}/sample`` — stage 3."""

    model_config = ConfigDict(extra="forbid")

    points: Annotated[list[DemAoiCorner], Field(min_length=1, max_length=5000)] = Field(
        ..., description="Points to sample, EPSG:4326."
    )
    names: list[str] | None = Field(
        None, description="Optional per-point labels; defaults to P1, P2, …"
    )
    method: Literal["bilinear", "nearest"] = "bilinear"
    offset_m: float = Field(0.0, description="Constant height offset added to every Z.")
    output_crs: str | None = Field(
        None, description="CRS for the returned X/Y. Defaults to the DEM's own CRS."
    )

    @field_validator("names")
    @classmethod
    def _names_match_points(cls, v: list[str] | None, info: Any) -> list[str] | None:
        points = info.data.get("points")
        if v is not None and points is not None and len(v) != len(points):
            raise ValueError(
                f"names has {len(v)} entries but points has {len(points)}; a label that "
                "silently pairs with the wrong coordinate is worse than no label"
            )
        return v


class DemSampleResponse(BaseModel):
    """``POST /dem/{run_id}/sample`` result."""

    model_config = ConfigDict(extra="forbid")

    run_id: str
    dem_crs: str = Field(..., description="The CRS the DEM was sampled in.")
    output_crs: str = Field(..., description="The CRS X/Y are reported in.")
    method: str
    offset_m: float
    points: list[DemSampledPoint]
    inside_count: int
    with_elevation_count: int


class DemActiveResponse(BaseModel):
    """``GET /dem/active`` — which DEM currently feeds GCP elevations."""

    model_config = ConfigDict(extra="allow")

    active: bool = Field(..., description="False when no DEM has been adopted.")
    provider_enabled: bool = Field(
        False,
        description=(
            "True when LE_ELEVATION_PROVIDER is `dem_run`. ★ A DEM can be adopted while "
            "the provider is off, in which case nothing samples it — reported rather "
            "than silently ignored."
        ),
    )
    path: str | None = None
    run_id: str | None = None
    source_name: str | None = None
    adopted_at: str | None = None
    output_crs: str | None = None
    output_pixel_size_m: list[float] | None = None
    vertical_ce90_m: float | None = Field(
        None, description="Reported on every GCP sampled from this DEM."
    )
    vertical_datum: str | None = Field(
        None,
        description=(
            "`unknown` unless declared. Copernicus/FABDEM are orthometric (EGM2008); "
            "GNSS altitudes are ellipsoidal and differ by ~20 m in Lebanon."
        ),
    )
    statistics: dict[str, Any] | None = None
